pkg/gc/gc_state_manager_test.go

```diff
@@ -0,0 +1,824 @@
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/pingcap/errors"
+
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/pkg/id"
+	"github.com/tikv/pd/pkg/keyspace"
+	"github.com/tikv/pd/pkg/mcs/utils/constant"
+	"github.com/tikv/pd/pkg/mock/mockcluster"
+	"github.com/tikv/pd/pkg/storage/endpoint"
+	"github.com/tikv/pd/pkg/storage/kv"
+	"github.com/tikv/pd/pkg/utils/etcdutil"
+	"github.com/tikv/pd/pkg/utils/keypath"
+	"github.com/tikv/pd/server/config"
+)
+
+type gcStateManagerTestSuite struct {
+	suite.Suite
+
+	storage  *endpoint.StorageEndpoint
+	provider endpoint.GCStateProvider
+	manager  *GCStateManager
+	clean    func()
+
+	keyspacePresets struct {
+		// A set of shortcuts for different kinds of keyspaces. Initialized in SetupTest.
+		// Tests are suggested to iterate over all these possibilities.
+
+		// All valid keyspaces.
+		all []uint32
+		// Subset of `all` that can manage their own GC. Includes NullKeyspace and keyspaces configured keyspace-level GC.
+		manageable []uint32
+		// all - manageable.
+		unmanageable []uint32
+		// Subset of `all` that uses unified GC (equals to unmanageable + NullKeyspace).
+		unifiedGC []uint32
+		// A set of not existing keyspace IDs. GC methods are mostly expected to fail on them.
+		notExisting []uint32
+		// A set of different keyspaceIDs that are expected to be regarded the same as NullKeyspaceID (0xffffffff).
+		// NullKeyspaceID is included.
+		nullSynonyms []uint32
+	}
+}
+
+func TestGCStateManager(t *testing.T) {
+	suite.Run(t, new(gcStateManagerTestSuite))
+}
+
+func newGCStateManagerForTest(t *testing.T) (storage *endpoint.StorageEndpoint, provider endpoint.GCStateProvider, gcStateManager *GCStateManager, clean func()) {
+	cfg := config.NewConfig()
+	re := require.New(t)
+
+	_, client, clean := etcdutil.NewTestEtcdCluster(t, 1)
+	kvBase := kv.NewEtcdKVBase(client)
+
+	// Simulate a member which id.Allocator may need to check.
+	err := kvBase.Save(keypath.LeaderPath(nil), "member1")
+	re.NoError(err)
+
+	s := endpoint.NewStorageEndpoint(kvBase, nil)
+	allocator := id.NewAllocator(&id.AllocatorParams{
+		Client: client,
+		Label:  id.KeyspaceLabel,
+		Member: "member1",
+		Step:   keyspace.AllocStep,
+	})
+	kgm := keyspace.NewKeyspaceGroupManager(context.Background(), s, client)
+	keyspaceManager := keyspace.NewKeyspaceManager(context.Background(), s, mockcluster.NewCluster(context.Background(), config.NewPersistOptions(cfg)), allocator, &config.KeyspaceConfig{}, kgm)
+	gcStateManager = NewGCStateManager(s.GetGCStateProvider(), cfg.PDServerCfg, keyspaceManager)
+
+	err = kgm.Bootstrap(context.Background())
+	re.NoError(err)
+	err = keyspaceManager.Bootstrap()
+	re.NoError(err)
+
+	// keyspaceID 0 exists automatically after bootstrapping.
+
+	ks1, err := keyspaceManager.CreateKeyspace(&keyspace.CreateKeyspaceRequest{
+		Name:       "ks1",
+		Config:     map[string]string{"gc_management_type": "global_gc"},
```

Comment:
it should be `unified_gc`  ?

Reply:
This should be updated when the renaming is done. As I see, the constant definitions in pkg/keyspace/keyspace.go is still not updated.
I'll add TODO comment here.